### ISSUE TITLE
js-interop: WebAssembly.instantiate() throws when bundling with `dart2js`

### ISSUE THREAD

**Issue by graphicbeacon** [dart-lang/sdk#36859]
Hi Dart Team,

As part of my ongoing investigation into Dart->JS interop, I have [an example with the WebAssembly API](https://creativebracket.com/how-to-load-a-webassembly-file-in-dart-for-web/) that seems to work when running the output generated with `dartdevc`.

Sadly I get this exception thrown in the browser after generating a `dart2js` output:

```
main.dart.js:133 Uncaught TypeError: a.au is not a function
    at es.$1 (main.dart.js:2829)
    at e1.O (main.dart.js:1975)
    at dW.$0 (main.dart.js:1896)
    at Object.ap (main.dart.js:823)
    at dO.$0 (main.dart.js:1858)
    at Object.h8 (main.dart.js:853)
    at dd.hc (main.dart.js:855)
    at dF.$1 (main.dart.js:1704)
    at hs (main.dart.js:208)
    at MutationObserver.<anonymous> (main.dart.js:217)
```
And the second error:
```
Uncaught (in promise) LinkError: WebAssembly.instantiate(): Import #0 module="imports" function="imported_func" error: function import requires a callable
```

**VM/Package details**

* Dart SDK version: **2.2.0**
* pkg/js version: **0.6.1+1**

**Failing code:**
https://gist.github.com/graphicbeacon/a2ed9b6e39a7391a893e8fe7dac25fa0

**Comment by sigmundch** [comment]
@graphicbeacon thanks for reaching out!

I have a hunch that the issue here might be that you may need to call `allowInterop` before passing a closure via js-interop.

DDC and dart2js have some discrepancies in how they handle js-interop (for example, https://github.com/dart-lang/sdk/issues/33134). We have a goal to unify them, but for now these differences can be hard to spot.

A very common one is handling of closures: DDC uses regular JavaScript closures and calling conventions that are similar to JavaScript. Dart2js does not, and to interoperate closures need to be wrapped before they are passed to JS.

Let us know if this addresses your issue. If not, try passing `-O1` to dart2js to make sure the code is compiled without minification, that will make the output and error message easier to debug.

**Comment by graphicbeacon** [comment]
Hey @sigmundch thanks for responding!

Got it working! The advice you gave really helped.

I had to use `allowInterop()` to wrap the closures. Disabling minification with `-O1` also helped debug another issue where the `.$then` method was invoked but was not defined. Resolved by creating a js-interop for the `Promise` API and using that.

I've been able to update my [post](https://creativebracket.com/how-to-load-a-webassembly-file-in-dart-for-web/) with the details.

Thanks so much.
